Userland/Applications/Maps/MapWidget.cpp

```diff
@@ -0,0 +1,180 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "MapWidget.h"
+#include <AK/URL.h>
+#include <LibGfx/ImageFormats/ImageDecoder.h>
+#include <LibProtocol/Request.h>
+
+// Math helpers
+static double radians(double degrees)
+{
+    return degrees * M_PI / 180.0;
+}
+
+static double degrees(double radians)
+{
+    return radians * 180.0 / M_PI;
+}
+
+// https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Pseudo-code
+static double longitude_to_tile_x(double longitude, int zoom)
+{
+    return pow(2, zoom) * ((longitude + 180.0) / 360.0);
+}
+
+static double latitude_to_tile_y(double latitude, int zoom)
+{
+    return pow(2, zoom) * (1.0 - (log(tan(radians(latitude)) + (1.0 / cos(radians(latitude)))) / M_PI)) / 2.0;
+}
+
+static double tile_x_to_longitude(double x, int zoom)
+{
+    return x / pow(2, zoom) * 360.0 - 180.0;
+}
+
+static double tile_y_to_latitude(double y, int zoom)
+{
+    return degrees(atan(sinh(M_PI * (1.0 - 2.0 * y / pow(2, zoom)))));
+}
+
+// MapWidget class
+MapWidget::MapWidget(Options const& options)
+    : m_tile_layer_url(options.tile_layer_url)
+    , m_center(options.center)
+    , m_zoom(options.zoom)
+{
+    m_request_client = Protocol::RequestClient::try_create().release_value_but_fixme_should_propagate_errors();
+}
+
+void MapWidget::mousedown_event(GUI::MouseEvent& event)
+{
+    if (event.button() == GUI::MouseButton::Primary) {
+        // Start map tiles dragging
+        m_dragging = true;
+        m_last_mouse_x = event.x();
+        m_last_mouse_y = event.y();
+        set_override_cursor(Gfx::StandardCursor::Drag);
+    }
+}
+
+void MapWidget::mousemove_event(GUI::MouseEvent& event)
+{
+    if (m_dragging) {
+        // Adjust map center by mouse delta
+        double delta_x = event.x() - m_last_mouse_x;
+        double delta_y = event.y() - m_last_mouse_y;
+        set_center({ tile_y_to_latitude(latitude_to_tile_y(m_center.latitude, m_zoom) - delta_y / TILE_SIZE, m_zoom),
+            tile_x_to_longitude(longitude_to_tile_x(m_center.longitude, m_zoom) - delta_x / TILE_SIZE, m_zoom) });
+        m_last_mouse_x = event.x();
+        m_last_mouse_y = event.y();
+        return;
+    }
+}
+
+void MapWidget::mouseup_event(GUI::MouseEvent&)
+{
+    // Stop map tiles dragging
+    if (m_dragging) {
+        m_dragging = false;
+        set_override_cursor(Gfx::StandardCursor::Arrow);
+        return;
+    }
+}
+
+void MapWidget::mousewheel_event(GUI::MouseEvent& event)
+{
+    if (event.wheel_delta_y() > 0) {
+        set_zoom(m_zoom - 1);
+    } else {
+        set_zoom(m_zoom + 1);
+    }
+}
+
+Optional<RefPtr<Gfx::Bitmap>> MapWidget::get_tile_image(int x, int y)
+{
+    // Get the right tile from tiles cache
+    for (Tile const& tile : m_tiles) {
+        if (tile.x == x && tile.y == y && tile.zoom == m_zoom) {
+            if (tile.image)
+                return tile.image;
+            return {};
+        }
+    }
```

Comment:
The main reason I didn't use a hashmap was because I could not write a hash function for the x y, zoom key struct. There is only `pair_int_hash` in `AK/HashFunctions.h`. And I didn't want to write my own custom hash function.